balancer-js/src/modules/pools/queries/params_builder.ts

```diff
@@ -0,0 +1,155 @@
+import * as PoolQueries from './types';
+import { Zero } from '@ethersproject/constants';
+import { getEncoder } from './get_encoder';
+
+export class ParamsBuilder implements PoolQueries.ParamsBuilder {
+  private encoder: PoolQueries.Encoder;
+
+  constructor(private pool: PoolQueries.Pool) {
+    const encoder = getEncoder(pool.poolType);
+    if (encoder) {
+      this.encoder = encoder;
+    } else {
+      throw 'Pool type not supported';
+    }
+  }
+
+  buildQueryJoinExactIn({
+    sender,
+    recipient = sender,
+    amountsIn,
+    minimumBPT = Zero,
+    fromInternalBalance = false,
+  }: PoolQueries.JoinExactInParams): PoolQueries.queryJoinParams {
+    const userData = this.encoder.joinExactTokensInForBPTOut(
+      amountsIn,
+      minimumBPT
+    );
+
+    const params = [
+      this.pool.id,
```

Comment:
About the `bind/this` discussion. Is this `this` that would point to the `pools module` instead of `queries` class?

Reply:
yes, `this` inside a class prototype doesn't need to be a class instance. it can refer to something completely different.